File: verifymate_changes/android/app/src/main/java/ai/journalmate/app/NotificationPlugin.java
package ai.journalmate.app;

import android.app.AlarmManager;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import androidx.core.app.ActivityCompat;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.JSObject;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Map;

@CapacitorPlugin(
    name = "NativeNotifications",
    permissions = {
        @Permission(
            strings = { "android.permission.POST_NOTIFICATIONS" },
            alias = "notifications"
        )
    }
)
public class NotificationPlugin extends Plugin {
    private static final String TAG = "NotificationPlugin";
    private static final String CHANNEL_ID = "journalmate_notifications";
    private static final String CHANNEL_NAME = "JournalMate Notifications";
    private static final String CHANNEL_DESC = "Notifications from JournalMate app";
    private static final String CHANNEL_ID_ALERTS = "journalmate_alerts";
    private static final String CHANNEL_NAME_ALERTS = "JournalMate Alerts";
    private static final String SCHEDULED_PREFS = "scheduled_notifications";

    @Override
    public void load() {
        super.load();
        createNotificationChannel();
        restoreScheduledNotifications();
    }

    /**
     * Restore scheduled notifications after app restart
     * This ensures alarms survive app force-close
     */
    private void restoreScheduledNotifications() {
        try {
            SharedPreferences prefs = getContext().getSharedPreferences(SCHEDULED_PREFS, Context.MODE_PRIVATE);
            Map<String, ?> all = prefs.getAll();
            int restored = 0;

            for (Map.Entry<String, ?> entry : all.entrySet()) {
                try {
                    JSONObject json = new JSONObject((String) entry.getValue());
                    long triggerAt = json.getLong("triggerAt");

                    if (triggerAt > System.currentTimeMillis()) {
                        // Future notification - reschedule
                        scheduleAlarm(
                            json.getInt("id"),
                            json.getString("title"),
                            json.getString("body"),
                            triggerAt
                        );
                        restored++;
                    } else {
                        // Past due - remove from storage
                        prefs.edit().remove(entry.getKey()).apply();
                    }
                } catch (JSONException e) {
                    Log.e(TAG, "Failed to parse scheduled notification: " + e.getMessage());
                    prefs.edit().remove(entry.getKey()).apply();
                }
            }

            if (restored > 0) {
                Log.d(TAG, "Restored " + restored + " scheduled notifications");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to restore scheduled notifications: " + e.getMessage());
        }
    }

    /**
     * Create notification channels (required for Android 8.0+)
     */
    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationManager notificationManager = getContext().getSystemService(NotificationManager.class);
            if (notificationManager == null) {
                Log.e(TAG, "NotificationManager is null, cannot create channels");
                return;
            }

            // Default channel for general notifications
            NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_DEFAULT
            );
            channel.setDescription(CHANNEL_DESC);
            channel.enableVibration(true);
            channel.setShowBadge(true);
            notificationManager.createNotificationChannel(channel);
            Log.d(TAG, "Default notification channel created: " + CHANNEL_ID);

            // HIGH importance channel for alerts and test notifications
            NotificationChannel alertChannel = new NotificationChannel(
                CHANNEL_ID_ALERTS,
                CHANNEL_NAME_ALERTS,
                NotificationManager.IMPORTANCE_HIGH
            );
            alertChannel.setDescription("Important alerts and test notifications");
            alertChannel.enableVibration(true);
            alertChannel.setVibrationPattern(new long[]{0, 500, 200, 500});
            alertChannel.setShowBadge(true);
            alertChannel.enableLights(true);
            notificationManager.createNotificationChannel(alertChannel);
            Log.d(TAG, "Alert notification channel created: " + CHANNEL_ID_ALERTS);
        }
    }

    /**
     * Check if notification permission is granted
     */
    @PluginMethod
    public void checkPermission(PluginCall call) {
        JSObject result = new JSObject();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // Android 13+ requires POST_NOTIFICATIONS permission
            boolean granted = ContextCompat.checkSelfPermission(
                getContext(),
                "android.permission.POST_NOTIFICATIONS"
            ) == PackageManager.PERMISSION_GRANTED;
            result.put("granted", granted);
        } else {
            // Before Android 13, notifications are enabled by default
            NotificationManagerCompat notificationManager = NotificationManagerCompat.from(getContext());
            result.put("granted", notificationManager.areNotificationsEnabled());
        }

        result.put("platform", "android");
        call.resolve(result);
    }

    /**
     * Request notification permission (Android 13+)
     */
    @PluginMethod
    public void requestPermission(PluginCall call) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // Android 13+ requires runtime permission
            if (ContextCompat.checkSelfPermission(
                    getContext(),
                    "android.permission.POST_NOTIFICATIONS"
                ) != PackageManager.PERMISSION_GRANTED) {

                // Request permission
                requestPermissionForAlias("notifications", call, "permissionCallback");
                return;
            }
        }

        // Permission already granted or not needed (pre-Android 13)
        JSObject result = new JSObject();
        result.put("granted", true);
        result.put("platform", "android");
        call.resolve(result);
    }

    /**
     * Callback after permission request
     */
    @PluginMethod
    public void permissionCallback(PluginCall call) {
        JSObject result = new JSObject();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            boolean granted = ContextCompat.checkSelfPermission(
                getContext(),
                "android.permission.POST_NOTIFICATIONS"
            ) == PackageManager.PERMISSION_GRANTED;
            result.put("granted", granted);
        } else {
            result.put("granted", true);
        }

        result.put("platform", "android");
        call.resolve(result);
    }

    /**
     * Show a local notification immediately
     * Uses HIGH importance alerts channel for visibility
     */
    @PluginMethod
    public void show(PluginCall call) {
        String title = call.getString("title", "JournalMate");
        String body = call.getString("body", "");
        Integer id = call.getInt("id", (int) System.currentTimeMillis());

        Log.d(TAG, "=== show() called ===");
        Log.d(TAG, "Title: " + title);
        Log.d(TAG, "Body: " + body);
        Log.d(TAG, "ID: " + id);

        try {
            // Create intent to open app when notification is tapped
            Intent intent = new Intent(getContext(), MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);

            PendingIntent pendingIntent = PendingIntent.getActivity(
                getContext(),
                id,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
            );

            // Build the notification using ALERTS channel for high visibility
            NotificationCompat.Builder builder = new NotificationCompat.Builder(getContext(), CHANNEL_ID_ALERTS)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(title)
                .setContentText(body)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(body))
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setCategory(NotificationCompat.CATEGORY_MESSAGE)
                .setAutoCancel(true)
                .setVibrate(new long[]{0, 500, 200, 500})
                .setContentIntent(pendingIntent);

            // Show the notification
            NotificationManagerCompat notificationManager = NotificationManagerCompat.from(getContext());

            // Check permission before showing (Android 13+)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                if (ContextCompat.checkSelfPermission(
                        getContext(),
                        "android.permission.POST_NOTIFICATIONS"
                    ) != PackageManager.PERMISSION_GRANTED) {
                    Log.w(TAG, "Notification permission not granted");
                    JSObject result = new JSObject();
                    result.put("success", false);
                    result.put("error", "Permission not granted. Please enable notifications in app settings.");
                    call.resolve(result);
                    return;
                }
            }

            notificationManager.notify(id, builder.build());
            Log.d(TAG, "=== Notification shown successfully with ID: " + id + " ===");

            JSObject result = new JSObject();
            result.put("success", true);
            result.put("id", id);
            call.resolve(result);

        } catch (Exception e) {
            Log.e(TAG, "Failed to show notification: " + e.getMessage());
            JSObject result = new JSObject();
            result.put("success", false);
            result.put("error", e.getMessage());
            call.resolve(result);
        }
    }

    /**
     * Cancel a notification by ID
     */
    @PluginMethod
    public void cancel(PluginCall call) {
        Integer id = call.getInt("id");

        if (id == null) {
            call.reject("Notification ID is required");
            return;
        }

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(getContext());
        notificationManager.cancel(id);

        JSObject result = new JSObject();
        result.put("success", true);
        call.resolve(result);
    }

    /**
     * Cancel all notifications
     */
    @PluginMethod
    public void cancelAll(PluginCall call) {
        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(getContext());
        notificationManager.cancelAll();

        // Also clear all scheduled notifications
        SharedPreferences prefs = getContext().getSharedPreferences(SCHEDULED_PREFS, Context.MODE_PRIVATE);
        prefs.edit().clear().apply();

        JSObject result = new JSObject();
        result.put("success", true);
        call.resolve(result);
    }

    /**
     * Schedule a notification for a future time
     * Uses AlarmManager for reliable delivery even when app is killed
     */
    @PluginMethod
    public void schedule(PluginCall call) {
        String title = call.getString("title", "JournalMate");
        String body = call.getString("body", "");
        Integer id = call.getInt("id", (int) System.currentTimeMillis());
        Long triggerAt = call.getLong("triggerAt");

        if (triggerAt == null) {
            call.reject("triggerAt is required (milliseconds since epoch)");
            return;
        }

        // Validate trigger time is in the future
        if (triggerAt <= System.currentTimeMillis()) {
            // If time is in the past, show immediately instead
            Log.w(TAG, "Scheduled time is in the past, showing immediately");
            call.getData().put("id", id);
            show(call);
            return;
        }

        Log.d(TAG, "Scheduling notification for " + triggerAt + ": " + title);

        try {
            // Schedule the alarm
            scheduleAlarm(id, title, body, triggerAt);

            // Store for restoration after app restart
            saveScheduledNotification(id, title, body, triggerAt);

            JSObject result = new JSObject();
            result.put("success", true);
            result.put("id", id);
            result.put("scheduledAt", triggerAt);
            call.resolve(result);

        } catch (Exception e) {
            Log.e(TAG, "Failed to schedule notification: " + e.getMessage());
            JSObject result = new JSObject();
            result.put("success", false);
            result.put("error", e.getMessage());
            call.resolve(result);
        }
    }

    /**
     * Get all pending scheduled notifications
     */
    @PluginMethod
    public void getPending(PluginCall call) {
        try {
            SharedPreferences prefs = getContext().getSharedPreferences(SCHEDULED_PREFS, Context.MODE_PRIVATE);
            Map<String, ?> all = prefs.getAll();

            JSObject result = new JSObject();
            JSObject notifications = new JSObject();

            for (Map.Entry<String, ?> entry : all.entrySet()) {
                try {
                    JSONObject json = new JSONObject((String) entry.getValue());
                    long triggerAt = json.getLong("triggerAt");

                    // Only include future notifications
                    if (triggerAt > System.currentTimeMillis()) {
                        JSObject notif = new JSObject();
                        notif.put("id", json.getInt("id"));
                        notif.put("title", json.getString("title"));
                        notif.put("body", json.getString("body"));
                        notif.put("triggerAt", triggerAt);
                        notifications.put(entry.getKey(), notif);
                    }
                } catch (JSONException e) {
                    Log.e(TAG, "Failed to parse notification: " + e.getMessage());
                }
            }

            result.put("notifications", notifications);
            result.put("count", notifications.length());
            call.resolve(result);

        } catch (Exception e) {
            call.reject("Failed to get pending notifications: " + e.getMessage());
        }
    }

    /**
     * Schedule an alarm using AlarmManager
     */
    private void scheduleAlarm(int id, String title, String body, long triggerAt) {
        Intent intent = new Intent(getContext(), NotificationAlarmReceiver.class);
        intent.setAction("SCHEDULED_NOTIFICATION");
        intent.putExtra("id", id);
        intent.putExtra("title", title);
        intent.putExtra("body", body);

        PendingIntent pendingIntent = PendingIntent.getBroadcast(
            getContext(),
            id,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
        );

        AlarmManager alarmManager = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);

        if (alarmManager == null) {
            Log.e(TAG, "AlarmManager not available");
            return;
        }

        // Use exact alarms when possible for precise timing
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            // Android 12+ - check if we can schedule exact alarms
            if (alarmManager.canScheduleExactAlarms()) {
                alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAt, pendingIntent);
                Log.d(TAG, "Scheduled exact alarm for ID " + id);
            } else {
                // Fall back to inexact alarm
                alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAt, pendingIntent);
                Log.d(TAG, "Scheduled inexact alarm for ID " + id + " (exact alarms not permitted)");
            }
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            // Android 6.0+ - use setExactAndAllowWhileIdle for Doze mode
            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAt, pendingIntent);
            Log.d(TAG, "Scheduled exact alarm for ID " + id);
        } else {
            // Older Android - use setExact
            alarmManager.setExact(AlarmManager.RTC_WAKEUP, triggerAt, pendingIntent);
            Log.d(TAG, "Scheduled alarm for ID " + id);
        }
    }

    /**
     * Save scheduled notification to SharedPreferences for restoration
     */
    private void saveScheduledNotification(int id, String title, String body, long triggerAt) {
        try {
            JSONObject json = new JSONObject();
            json.put("id", id);
            json.put("title", title);
            json.put("body", body);
            json.put("triggerAt", triggerAt);

            SharedPreferences prefs = getContext().getSharedPreferences(SCHEDULED_PREFS, Context.MODE_PRIVATE);
            prefs.edit().putString(String.valueOf(id), json.toString()).apply();

            Log.d(TAG, "Saved scheduled notification: " + id);
        } catch (JSONException e) {
            Log.e(TAG, "Failed to save scheduled notification: " + e.getMessage());
        }
    }

    /**
     * Remove a scheduled notification from storage
     * Called when notification is shown or cancelled
     */
    public static void removeScheduledNotification(Context context, int id) {
        SharedPreferences prefs = context.getSharedPreferences(SCHEDULED_PREFS, Context.MODE_PRIVATE);
        prefs.edit().remove(String.valueOf(id)).apply();
        Log.d(TAG, "Removed scheduled notification: " + id);
    }
}

File: ios/App/App/NotificationPlugin.swift
import Foundation
import Capacitor
import UserNotifications
import UIKit

@objc(NativeNotifications)
public class NotificationPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeNotifications"
    public let jsName = "NativeNotifications"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "schedule", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancelAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPending", returnType: CAPPluginReturnPromise)
    ]

    private let scheduledNotificationsKey = "scheduled_notifications"

    public override func load() {
        super.load()
        print("[NotificationPlugin] Plugin loaded")
        // Restore scheduled notifications on load (handles app restart)
        restoreScheduledNotifications()
    }

    // MARK: - Check Permission

    @objc func checkPermission(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let granted = settings.authorizationStatus == .authorized ||
                          settings.authorizationStatus == .provisional

            call.resolve([
                "granted": granted,
                "platform": "ios",
                "status": self.authStatusToString(settings.authorizationStatus)
            ])
        }
    }

    // MARK: - Request Permission

    @objc func requestPermission(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
            if let error = error {
                print("[NotificationPlugin] Permission error: \(error.localizedDescription)")
            }

            call.resolve([
                "granted": granted,
                "platform": "ios"
            ])
        }
    }

    // MARK: - Show Immediate Notification

    @objc func show(_ call: CAPPluginCall) {
        let title = call.getString("title") ?? "JournalMate"
        let body = call.getString("body") ?? ""
        let id = call.getInt("id") ?? Int(Date().timeIntervalSince1970)

        print("[NotificationPlugin] show() called - Title: \(title), Body: \(body), ID: \(id)")

        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        content.badge = 1

        // Add category for action buttons if specified
        if let category = call.getString("category") {
            content.categoryIdentifier = category
        }

        // Add userInfo for deep linking
        var userInfo: [String: Any] = [
            "notificationId": id
        ]
        if let route = call.getString("route") {
            userInfo["route"] = route
        }
        content.userInfo = userInfo

        // Trigger immediately (1 second delay to ensure delivery)
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)

        let request = UNNotificationRequest(
            identifier: "notification_\(id)",
            content: content,
            trigger: trigger
        )

        UNUserNotificationCenter.current().add(request) { error in
            if let error = error {
                print("[NotificationPlugin] Failed to show notification: \(error.localizedDescription)")
                call.resolve([
                    "success": false,
                    "error": error.localizedDescription
                ])
            } else {
                print("[NotificationPlugin] Notification shown successfully with ID: \(id)")
                call.resolve([
                    "success": true,
                    "id": id
                ])
            }
        }
    }

    // MARK: - Schedule Notification

    @objc func schedule(_ call: CAPPluginCall) {
        let title = call.getString("title") ?? "JournalMate"
        let body = call.getString("body") ?? ""
        let id = call.getInt("id") ?? Int(Date().timeIntervalSince1970)

        guard let triggerAt = call.getDouble("triggerAt") else {
            call.reject("triggerAt is required (milliseconds since epoch)")
            return
        }

        let triggerDate = Date(timeIntervalSince1970: triggerAt / 1000.0)

        // If time is in the past, show immediately
        if triggerDate <= Date() {
            print("[NotificationPlugin] Scheduled time is in the past, showing immediately")
            show(call)
            return
        }

        print("[NotificationPlugin] Scheduling notification for \(triggerDate): \(title)")

        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        content.badge = 1

        // Add category for action buttons
        if let category = call.getString("category") {
            content.categoryIdentifier = category
        } else {
            content.categoryIdentifier = "TASK_REMINDER"
        }

        // Add userInfo for deep linking
        var userInfo: [String: Any] = [
            "notificationId": id,
            "title": title,
            "body": body,
            "triggerAt": triggerAt
        ]
        if let route = call.getString("route") {
            userInfo["route"] = route
        }
        if let taskId = call.getString("taskId") {
            userInfo["taskId"] = taskId
        }
        content.userInfo = userInfo

        // Create calendar-based trigger
        let triggerComponents = Calendar.current.dateComponents(
            [.year, .month, .day, .hour, .minute, .second],
            from: triggerDate
        )
        let trigger = UNCalendarNotificationTrigger(dateMatching: triggerComponents, repeats: false)

        let request = UNNotificationRequest(
            identifier: "notification_\(id)",
            content: content,
            trigger: trigger
        )

        UNUserNotificationCenter.current().add(request) { error in
            if let error = error {
                print("[NotificationPlugin] Failed to schedule notification: \(error.localizedDescription)")
                call.resolve([
                    "success": false,
                    "error": error.localizedDescription
                ])
            } else {
                print("[NotificationPlugin] Notification scheduled successfully for ID: \(id)")

                // Save to UserDefaults for restoration
                self.saveScheduledNotification(id: id, title: title, body: body, triggerAt: triggerAt)

                call.resolve([
                    "success": true,
                    "id": id,
                    "scheduledAt": triggerAt
                ])
            }
        }
    }

    // MARK: - Cancel Notification

    @objc func cancel(_ call: CAPPluginCall) {
        guard let id = call.getInt("id") else {
            call.reject("Notification ID is required")
            return
        }

        let identifier = "notification_\(id)"
        UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: [identifier])
        UNUserNotificationCenter.current().removeDeliveredNotifications(withIdentifiers: [identifier])

        // Remove from saved notifications
        removeScheduledNotification(id: id)

        print("[NotificationPlugin] Cancelled notification with ID: \(id)")
        call.resolve([
            "success": true
        ])
    }

    // MARK: - Cancel All Notifications

    @objc func cancelAll(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
        UNUserNotificationCenter.current().removeAllDeliveredNotifications()

        // Clear saved notifications
        UserDefaults.standard.removeObject(forKey: scheduledNotificationsKey)

        print("[NotificationPlugin] Cancelled all notifications")
        call.resolve([
            "success": true
        ])
    }

    // MARK: - Get Pending Notifications

    @objc func getPending(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().getPendingNotificationRequests { requests in
            var notifications: [[String: Any]] = []

            for request in requests {
                var notif: [String: Any] = [
                    "identifier": request.identifier,
                    "title": request.content.title,
                    "body": request.content.body
                ]

                if let trigger = request.trigger as? UNCalendarNotificationTrigger,
                   let nextDate = trigger.nextTriggerDate() {
                    notif["triggerAt"] = nextDate.timeIntervalSince1970 * 1000.0
                }

                notifications.append(notif)
            }

            call.resolve([
                "notifications": notifications,
                "count": notifications.count
            ])
        }
    }

    // MARK: - Helper Methods

    private func authStatusToString(_ status: UNAuthorizationStatus) -> String {
        switch status {
        case .notDetermined:
            return "notDetermined"
        case .denied:
            return "denied"
        case .authorized:
            return "authorized"
        case .provisional:
            return "provisional"
        case .ephemeral:
            return "ephemeral"
        @unknown default:
            return "unknown"
        }
    }

    private func saveScheduledNotification(id: Int, title: String, body: String, triggerAt: Double) {
        var saved = UserDefaults.standard.dictionary(forKey: scheduledNotificationsKey) as? [String: [String: Any]] ?? [:]

        saved[String(id)] = [
            "id": id,
            "title": title,
            "body": body,
            "triggerAt": triggerAt
        ]

        UserDefaults.standard.set(saved, forKey: scheduledNotificationsKey)
        print("[NotificationPlugin] Saved scheduled notification: \(id)")
    }

    private func removeScheduledNotification(id: Int) {
        var saved = UserDefaults.standard.dictionary(forKey: scheduledNotificationsKey) as? [String: [String: Any]] ?? [:]
        saved.removeValue(forKey: String(id))
        UserDefaults.standard.set(saved, forKey: scheduledNotificationsKey)
        print("[NotificationPlugin] Removed scheduled notification: \(id)")
    }

    private func restoreScheduledNotifications() {
        guard let saved = UserDefaults.standard.dictionary(forKey: scheduledNotificationsKey) as? [String: [String: Any]] else {
            return
        }

        let currentTime = Date().timeIntervalSince1970 * 1000.0
        var restored = 0

        for (key, data) in saved {
            guard let id = data["id"] as? Int,
                  let title = data["title"] as? String,
                  let body = data["body"] as? String,
                  let triggerAt = data["triggerAt"] as? Double else {
                continue
            }

            if triggerAt > currentTime {
                // Future notification - reschedule
                let content = UNMutableNotificationContent()
                content.title = title
                content.body = body
                content.sound = .default
                content.categoryIdentifier = "TASK_REMINDER"

                let triggerDate = Date(timeIntervalSince1970: triggerAt / 1000.0)
                let triggerComponents = Calendar.current.dateComponents(
                    [.year, .month, .day, .hour, .minute, .second],
                    from: triggerDate
                )
                let trigger = UNCalendarNotificationTrigger(dateMatching: triggerComponents, repeats: false)

                let request = UNNotificationRequest(
                    identifier: "notification_\(id)",
                    content: content,
                    trigger: trigger
                )

                UNUserNotificationCenter.current().add(request) { error in
                    if let error = error {
                        print("[NotificationPlugin] Failed to restore notification \(id): \(error)")
                    }
                }
                restored += 1
            } else {
                // Past due - remove from storage
                removeScheduledNotification(id: id)
            }
        }

        if restored > 0 {
            print("[NotificationPlugin] Restored \(restored) scheduled notifications")
        }
    }
}
